### PR TITLE
[NPM-3239] Make local resolver address and pid to container caches persistent across updates

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Makes the two caches used for local address resolution in the process-agent persistent across cache updates. Previously we would reset the caches to the latest containers update. This can potentially remove address/pid info for a container that could still have its connections in the connections check.

The caches now use an in use marker for their entries. On every containers update (every 10s), all entries are marked as not in use initially and subsequently only the entries from the update are marked as in use. We keep the not in use entries as they can potentially still be needed by an upcoming connections check (during a call to `LocalResolver.Resolve`). The not in use entries are flushed by `LocalResolver.Resolve` (every `30s` when the connections check runs).

The performance impact on process-agent in load testing has been minimal; this will be monitored as it is rolled out to staging for any perf impact.

### CODE REVIEW

**Review by leeavital — COMMENTED** [review]
It seems like this is written with the implicit assumption that there will be two container-pull cycles (calls to LoadContainers) per call to Resolve. 

Given that, did you consider having LoadContainer always "merge" the current containers/pids instead of clearing addrToCtrID and   ctrForPid during each call the LoadContainers ?

**Comment by hmahmood** [comment]
> It seems like this is written with the implicit assumption that there will be two container-pull cycles (calls to LoadContainers) per call to Resolve.
> 
> Given that, did you consider having LoadContainer always "merge" the current containers/pids instead of clearing addrToCtrID and ctrForPid during each call the LoadContainers ?

That's what this we are doing here essentially, but there has to be a way to clear out entries that are no longer needed. Like containers that do not exist or pids that do not exist.

**Comment by hmahmood** [comment]
> > That's what this we are doing here essentially, but there has to be a way to clear out entries that are no longer needed. Like containers that do not exist or pids that do not exist anymore.
> 
> I'm proposing that each call to Resolve() do a full clear of the maps at the end. As long as we're guaranteed to have a call to LoadContainers() before the next call to Resolve(), Resolve() should be operating on up to date data.

The next call to LoadContainers may not have some of pids/containers that we discarded in the last call to Resolve.

**Comment by leeavital** [comment]
> The next call to LoadContainers may not have some of pids/containers that we discarded in the last call to Resolve

Aren't we guaranteed to have a call to LoadContainers() which has a snapshot of all containers before the next call to resolve?

It might be helpful to chat synchronously.

File: pkg/process/checks/net.go [hunk 1]
```diff
@@ -32,6 +32,11 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/subscriptions"
 )
 
+const (
+	maxResolverPidCacheSize  = 32768
```
Comment (akarpz) [inline pkg/process/checks/net.go:36]: how were these sizes chosen? it doesn't look like there was a limit before

**Review by akarpz — APPROVED** [review]

**Review by daniel-taf — APPROVED** [review]
One general open question to help my understanding: Do the connections pulled from the sys-probe represent all connections seen within the last 30 seconds, active or not?

Thus, if a connection is short-lived, the address of the container can still be resolved if needed with this change.

File: pkg/process/checks/net.go [hunk 2]
```diff
@@ -32,6 +32,11 @@ import (
...
+	maxResolverAddrCacheSize = 4096
+)
```
[the remark this answers is not in the document]
Comment (hmahmood) [reply pkg/process/checks/net.go:38]: @akarpz had the same question.

These are more of a precaution than anything else. They probably will not work for some environments, but I was not comfortable with not putting an upper limit on the number of entries in the caches.

[the remark this answers is not in the document]
Comment (hmahmood) [reply pkg/process/checks/net.go:38]: There are warn logs put out when we drop: https://github.com/DataDog/datadog-agent/pull/24122/files#diff-5960d03d261e9df331687c41970746f68fce1002ef37909f292528e292830d37R115 https://github.com/DataDog/datadog-agent/pull/24122/files#diff-5960d03d261e9df331687c41970746f68fce1002ef37909f292528e292830d37R135

**Pull request merged**